# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [4]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct!

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*At just glancing at the dataset logistic Regression might perform better because of binary nature of values present in the labled column spam.*

## Split the Data into Training and Testing Sets

In [5]:
# Create the labels set `y` and features DataFrame `X`
y = data["spam"]
X = data.copy().drop(columns="spam")

In [6]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [7]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
4576            0.00                0.0           0.00           0.0   
4401            0.00                0.0           0.00           0.0   
3707            0.17                0.0           0.17           0.0   
2362            0.00                0.0           0.00           0.0   
1537            0.00                0.0           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
4576           0.00             0.0               0.0                 0.0   
4401           0.00             0.0               0.0                 0.0   
3707           0.00             0.0               0.0                 0.0   
2362           0.00             0.0               0.0                 0.0   
1537           2.17             0.0               0.0                 0.0   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
4576              0.0             0.0  ...                   0.0        0.000   
4401              0.0             0.0  ...                   0.0        0.000   
3707              0.8             0.0  ...                   0.0        0.253   
2362              0.0             0.0  ...                   0.0        0.000   
1537              0.0             0.0  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
4576        0.131        0.000          0.0        0.000          0.0   
4401        0.000        0.000          0.0        0.000          0.0   
3707        0.168        0.084          0.0        0.024          0.0   
2362        0.000        0.000          0.0        0.000          0.0   
1537        0.000        0.000          0.0        0.000          0.0   

      capital_run_length_average  capital_run_length_longest  \
4576                       1.488                           5   
4401                       1.571                           5   
3707                       4.665                          81   
2362                       4.228                          53   
1537                       1.333                           5   

      capital_run_length_total  
4576                        64  
4401                        11  
3707                      1031  
2362                       148  
1537                        16  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [8]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [10]:
# Fit the Standard Scaler with the training data
scaler.fit(X_train)

StandardScaler()

In [12]:
# Scale the training data
X_train_scaled = scaler.transform(X_train)
X_train_scaled[:1]

array([[-0.35811925, -0.16744248, -0.56071981, -0.04464735, -0.45989286,
        -0.34953634, -0.28922113, -0.25739046, -0.32549439, -0.36823638,
        -0.29678298, -0.20093626,  0.89610632, -0.17814177, -0.19548577,
        -0.2908286 , -0.3143283 , -0.3590312 , -0.11026558, -0.16783359,
        -0.67410557, -0.12560995, -0.30055631, -0.2115886 , -0.3195345 ,
        -0.29682742, -0.22880684, -0.22088119, -0.16391038, -0.21448971,
        -0.16694474, -0.13976611, -0.16993983, -0.14126662, -0.18103904,
         0.66617163, -0.32175894, -0.06173808, -0.17869623, -0.18133233,
         0.95960199, -0.17292191, -0.20437083, -0.13855365,  0.79403889,
         0.23503906, -0.07263589, -0.10512187, -0.15710024, -0.02804402,
        -0.14912157, -0.31596957, -0.30993787, -0.10477568, -0.13974678,
        -0.23023328, -0.37548424]])

In [13]:
# Scale the test data
X_test_scaled = scaler.transform(X_test)
X_test_scaled[:1]

array([[-0.35811925, -0.16744248, -0.56071981, -0.04464735,  1.040797  ,
        -0.34953634, -0.28922113, -0.25739046, -0.32549439, -0.36823638,
        -0.29678298, -0.62434926, -0.31249394, -0.17814177, -0.19548577,
        -0.2908286 , -0.3143283 , -0.3590312 , -0.9354155 , -0.16783359,
        -0.67410557, -0.12560995, -0.30055631, -0.2115886 , -0.3195345 ,
        -0.29682742, -0.22880684, -0.22088119, -0.16391038, -0.21448971,
        -0.16694474, -0.13976611, -0.16993983, -0.14126662, -0.18103904,
        -0.23296566, 11.62259811, -0.06173808, -0.17869623, -0.18133233,
        -0.11906298, -0.17292191, -0.20437083, -0.13855365, -0.30435906,
        -0.20940641, -0.07263589, -0.10512187, -0.15710024, -0.17564676,
        -0.14912157, -0.31596957,  0.04546352, -0.10477568,  0.07078032,
        -0.09748408, -0.12139738]])

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`.

In [14]:
# Create and fit a Logistic Regression model
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(random_state=1)
lr.fit(X_train_scaled, y_train)

LogisticRegression(random_state=1)

In [15]:
# Make and save testing predictions with the trained Logistic Regression model using the test data
testing_predictions = lr.predict(X_test_scaled)

# Review the predictions
testing_predictions

array([0, 0, 1, ..., 0, 0, 1])

In [16]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predictions)

0.9278887923544744

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`.

In [18]:
# Create and fit a Random Forest Classifier model
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(random_state=1)
rfc.fit(X_train_scaled, y_train)

RandomForestClassifier(random_state=1)

In [19]:
# Make and save testing predictions with the trained Random Forest Classifier model using the test data
testing_predictions = rfc.predict(X_test_scaled)

# Review the predictions
testing_predictions

array([1, 1, 1, ..., 1, 0, 1])

In [20]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predictions)

0.9669852302345786

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

*The accuracy score from RandomForestClassifier is 0.9669852302345786 which looks better than the LogisticRegression accuracy score of 0.9278887923544744. Based on the score RandomForestClassifier performed better than LogisticRegression, proving my initial guess is wrong.

The reasons might be because there is an imbalance in y value counts, missing values in the dataset or large number of features.*